components/ai_chat/core/browser/engine/engine_consumer_conversation_api.cc

```diff
@@ -202,8 +207,73 @@ void EngineConsumerConversationAPI::GenerateAssistantResponse(
     conversation.push_back(
         GetAssociatedContentConversationEvent(page_content, is_video));
   }
-  // history
-  for (const auto& message : conversation_history) {
+
+  // We're going to iterate over the conversation entries and
+  // build a list of events for the remote API.
+  // We largely want to send the full conversation with all events back to the
+  // model in order to preserve the context of the conversation. However, some
+  // tool results are extremely large (especially for images), and repetitive.
+  // We need a way to remove the noise in order to 1) not overwhelm the model
+  // and 2) not surpass the max token limit. For now, this is a rudimentary
+  // approach that only keeps the most recent large tool results. Use a two-pass
+  // approach: first identify which large tool results to keep, then build the
+  // conversation in chronological order.
+
+  // Step 1: First pass - identify large tool results and remember which ones to
+  // remove.
+  absl::flat_hash_set<std::pair<size_t, size_t>> large_tool_result_remove_set;
+  size_t large_tool_count = 0;
+
+  for (size_t message_index = conversation_history.size(); message_index > 0;
+       --message_index) {
```

Comment:
hey I think 
```suggestion
  for (size_t message_index = conversation_history.size() - 1; message_index >= 0;
       --message_index) {
```
is probably more idiomatic (so the index isn't off by one)

Reply:
The problem with this is it causes a crash when size_t wraps around. At least I think that's what was happening - happy to take obvious advice I'm missing.